### DISCUSSION TITLE
TCA migration questions and some experiences from 0.54 to 1.0

### DISCUSSION THREAD

**Discussion by marcuswu0814** [pointfreeco/swift-composable-architecture#2345] [General]
If someone like me wanna find some guide to migrate to `1.0` version, it’s some of my experiences when I adapt to my projects.

Also had some questions in sections if someone can share your advices, really thanks for that! 

## 3. Using closure base api for init a `Store`

From: 
```swift
ContentView(
    store: Store(
        initialState: AppFeature.State(),
        reducer: AppFeature() // <- ‼️ Remove this 
    )
)
```

To:
```swift
ContentView(
    store: Store(
        initialState: AppFeature.State()
    ) { // <- Trailing closure with Reducer here 
        AppFeature()
    }
)
```
> **important**
>  Question: But why? Any benefit for this change?

## 4. Sending action to `Store` immediately
In past time we need to send action to a `ViewStore` although you didn’t need observe any changes.  

In 1.0, you just need to send the action to `Store`.

From:
```swift
Button("Some Button") {
    ViewStore(store).send(.delegate(.tap))
}
```

To:
```swift
Button("Some Button") {
    store.send(.delegate(.tap))
}
```

## 5. @BindableState to @BindingState

Replace all `@BindableState` to `@BindingState` in your `State` struct.

In your view, your need to change binding syntax

From:
```swift
TextEditor(text: store.binding(\.$text))
```

To:
```swift
TextEditor(text: store.$text)
```

> **important**
> Question: In past, I used `ViewStore(store).binding(\.$text)` in some where I didn’t have completely  `ViewStore` causing by `WithViewStore(...)`.
> 
> I know there is a [guide](https://github.com/pointfreeco/swift-composable-architecture/discussions/2216) for this situation but for me it’s too many boilerplate `ViewState`  code need to write and hope can be solve by Swift Macro soon.
> 
> Is this the only way I can do? Or can I still using `ViewStore(store, observe: { $0 }).$text` in my view?

## 6. Loose coupling of `Combine`

Lots of feature be replace with Swift Concurrency with `async/await`, if your API is still using `Publisher`, you can follow the [tip](https://github.com/pointfreeco/swift-composable-architecture/discussions/1987#discussioncomment-5384255) and it still work, if you need the `eraseToEffect()` or `catchToEffect()`, just adding the extension back for your situation.

In my case, my API using `RxSwift`’s `Single` trait and I using the follow extension work with `Effect`:

```swift
public extension Single where Trait == SingleTrait {
    func catchToEffect<T>(
      _ transform: @escaping (Result<Element, Error>) -> T
    ) -> Effect<T> {
        .publisher {
            publisher.map(
                withEscapedDependencies { escaped in
                    { action in
                        escaped.yield {
                            transform(.success(action))
                        }
                    }
                }
            )
            .catch { Just(transform(.failure($0))) }
        }
    }
}
```

The `publisher` property of `Single` if from [RxCombine](https://github.com/CombineCommunity/RxCombine).

---

I’m not sure is all of my change is best practice or correctly, if there had any problems, please feel free to let me know, and also glad to your sharing!

**Comment by mbrandonw** [comment]
Hi @marcuswu0814, thank you for writing this up to share with everyone.

I will say, though, that we highly recommended people upgrade to 0.58.0 _before_ going to 1.0. That would have helped catch most of these things as simpler deprecation warnings rather than hard, incomprehensible errors.
